harness/determined/pytorch/_pytorch_context.py

```diff
@@ -57,4 +82,224 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+            if lr_scheduler is not None:
+                self.lr_schedulers.append(lr_scheduler)
+
+            amp_setting = (
+                self.get_experiment_config().get("optimizations", {}).get("mixed_precision", None)
+            )
+            if amp_setting is not None:
+                self._configure_amp(model, optim, amp_setting)
+
+        self.callbacks = trial.build_callbacks()
+
+        # Since trial is instantiated outside this function, we will only check if
+        # the model and the optimizer is initialized.
+        check.is_not_none(self.model, "Must have one model")
+        check.gt_eq(len(self.optimizers), 1, "Must have at least one optimizer")
+
+    def _Model(self, model: torch.nn.Module) -> torch.nn.Module:
+        self.model = model.to(self.device)
+        if not self.hvd_config.use and self.n_gpus > 1:
+            check.eq(
+                self.hvd_config.aggregation_frequency,
+                1,
+                "Please enable `optimized_parallel` to use aggregation "
+                "frequency greater than 1 for single machine multi-GPU "
+                "training.",
+            )
+            self.model = nn.DataParallel(self.model)
+            logging.debug("Initialized mode for native parallel training.")
+        return model
+
+    def _Optimizer(self, optimizer: torch.optim.Optimizer) -> torch.optim.Optimizer:  # type: ignore
+        check.is_not_none(self.model, "Must have one model first.")
+
+        # TODO: Check that optimizer is not an amp optimizer.
+        if self.hvd_config.use:
+
+            def filter_named_parameters(
+                model: nn.Module, optimizer: torch.optim.Optimizer,  # type: ignore
+            ) -> List:
+                """
+                filter_named_parameters filters the named parameters of a specified optimizer out
+                of all the named parameters from a specified model. We need this function because
+                a torch.optim.Optimizer doesn't store parameter names and we need the names of
+                the parameters when mapping parameters to each horovod.DistributedOptimizer.
+                """
+                opt_params = {
+                    p for group in optimizer.param_groups for p in group.get("params", [])
+                }
+                return [(name, p) for name, p in model.named_parameters() if p in opt_params]
+
+            use_compression = self.hvd_config.fp16_compression
+            optimizer = hvd.DistributedOptimizer(
+                optimizer,
+                named_parameters=filter_named_parameters(cast(nn.Module, self.model), optimizer),
+                backward_passes_per_step=self.hvd_config.aggregation_frequency,
+                compression=hvd.Compression.fp16 if use_compression else hvd.Compression.none,
+            )
+            logging.debug("Initialized optimizer for distributed and optimized parallel training.")
+
+        self.optimizers.append(optimizer)
+        return optimizer
```

Comment:
We should consider modifying their optimizer to fail if they call `optimizer.zero_grad()` themselves, or ignore it but issue a warning or something.

Reply:
Yes. I thought so, too. However, we should create a new class for optimizers as opposed to monkey patching the optimizers.

Reply:
Horovod already creates a class dynamically around the original optimizer, so I am not confident that it makes a lot of difference here.